File: src/com/haskforce/cabal/highlighting/CabalAnnotator.java
package com.haskforce.cabal.highlighting;

import com.haskforce.cabal.psi.CabalTypes;
import com.intellij.lang.annotation.AnnotationHolder;
import com.intellij.lang.annotation.Annotator;
import com.intellij.openapi.editor.colors.EditorColorsManager;
import com.intellij.openapi.editor.colors.TextAttributesKey;
import com.intellij.psi.PsiElement;
import com.intellij.psi.impl.source.tree.LeafPsiElement;
import org.jetbrains.annotations.NotNull;

public class CabalAnnotator implements Annotator {
    @Override
    public void annotate(@NotNull PsiElement element, @NotNull AnnotationHolder holder) {
        if (element instanceof LeafPsiElement) {
            LeafPsiElement leaf = (LeafPsiElement)element;
            // Fix highlighting of keys where jflex falls short.
            if (leaf.getElementType() == CabalTypes.SIMPLEKEY) {
                setHighlighting(element, holder, CabalSyntaxHighlighter.KEY);
            }
        }
    }

    private static void setHighlighting(@NotNull PsiElement element, @NotNull AnnotationHolder holder,
                                        @NotNull TextAttributesKey key) {
        holder.createInfoAnnotation(element, null).setEnforcedTextAttributes(
                EditorColorsManager.getInstance().getGlobalScheme().getAttributes(key));
    }
}

File: src/com/haskforce/cabal/highlighting/CabalAnnotator.scala
package com.haskforce.cabal.highlighting

import com.intellij.lang.annotation.{AnnotationHolder, Annotator}
import com.intellij.openapi.editor.colors.{EditorColorsManager, TextAttributesKey}
import com.intellij.psi.PsiElement
import com.haskforce.cabal.lang.psi._
import com.haskforce.cabal.lang.psi.CabalTypes._

class CabalAnnotator extends Annotator {

  def annotate(el: PsiElement, h: AnnotationHolder): Unit = {
    el.getNode.getElementType match {
      case _: CabalFieldKeyTokenType => setHighlighting(el, h, CabalSyntaxHighlighter.KEY)
      case _: CabalStanzaKeyTokenType => setHighlighting(el, h, CabalSyntaxHighlighter.CONFIG)
      case _: CabalStanzaArgTokenType => setHighlighting(el, h, CabalSyntaxHighlighter.CONFIG)
      case LBRACE | RBRACE => setHighlighting(el, h, CabalSyntaxHighlighter.COLON)
      case _ => // noop
    }
  }

  private def setHighlighting(element: PsiElement, holder: AnnotationHolder, key: TextAttributesKey) {
    holder
        .createWeakWarningAnnotation(element, "Message")
      .setEnforcedTextAttributes(EditorColorsManager.getInstance.getGlobalScheme.getAttributes(key))
  }
}
